types/searchResult.js

```diff
@@ -0,0 +1,70 @@
+const EAKInfo = require('./eakInfo')
+const Meta = require('./meta')
+const Record = require('./record')
+
+/**
+ * Describe a query result returned from E3DB API.
+ */
+class SearchResult {
+  constructor(client, searchRequest) {
+    this.client = client
+    this.request = searchRequest
+    this.done = false
+    this.totalResults = -1
+  }
+
+  /**
+   * Get the next page of results from the current query
+   *
+   * @returns {Promise<array>}
+   */
+  async next() {
+    // Finished iteration, exit early
+    if (this.done) {
+      return []
+    }
+
+    let response = await this.client._search(this.request)
+    // If we've reached the last page, keep track and exit
+    if (response.last_index === 0) {
+      this.done = true
+    }
+
+    if (!response.results) {
+      return []
+    }
+
+    /* eslint-disable */
+    let records = await Promise.all(
+      response.results.map(async result => {
+        const meta = await Meta.decode(result.meta)
+        const record = new Record(meta, result.record_data)
+        if (this.request.includeData) {
+        }
+        if (this.request.includeData && result.access_key !== null) {
+          const eak = await EAKInfo.decode(result.access_key)
+          const ak = await this.client.crypto.decryptEak(
+            this.client.config.privateKey,
+            eak
+          )
+          return this.client.crypto.decryptRecord(record, ak)
+        }
+
+        return record
+      })
+    )
+
+    this.request.nextToken = response.last_index
+    this.totalResults = response.total_results
+    if (response.search_id) {
+      this.searchId = response.search_id
+    }
+    if (records.length < this.request.count) {
```

Comment:
I *believe* because search service is eventually consistent this could lead to a situation where an un-share happens at search time where PDS will only return a subset of records returns from the ES search.  

The above last_index determination of done is more reliable, so I don't think this check is necessary
```
    if (response.last_index === 0) {
      this.done = true
    }
```

Reply:
I'll look at it, but I've run into bugs with poor server reporting of done -- this fixes it, but it is a bandaid. That was mostly with search v1, too.

Reply:
OK -- here is my question for search V2 -- when you reach the last page of results, do you have to make a request that is out of range before you get back a last_index of 0?

which is correct?:

request next null:
  return 10 results and last index of 9
request next 9:
  return 4 results and last index of 13
request next 13:
  return 0 results and last index of 0

== OR ==

request next null:
  return 10 results and last index of 9
request next 9:
  return 4 results and last index of 0